src/components/CurrentUserProvider.js

```diff
@@ -0,0 +1,151 @@
+import { useDataQuery } from '@dhis2/app-runtime'
+import i18n from '@dhis2/d2-i18n'
+import { Layer, CenteredContent, CircularLoader, NoticeBox } from '@dhis2/ui'
+import PropTypes from 'prop-types'
+import React, { createContext } from 'react'
+
+const CurrentUserContext = createContext({})
+
+const query = {
+    me: {
+        resource: '/me',
+        params: {
+            fields: [
+                'id',
+                'username',
+                'surname',
+                'firstName',
+                'authorities',
+                'userCredentials[userRoles[id]]',
+                'userGroups[id]',
+                'organisationUnits[id,path,displayName,children::isNotEmpty]',
+                'dataViewOrganisationUnits[id,path,displayName,children::isNotEmpty]',
+                'teiSearchOrganisationUnits[id,path,displayName,children::isNotEmpty]',
+            ],
+        },
+    },
+    systemOrganisationUnitRoots: {
+        resource: '/organisationUnits',
+        params: {
+            paging: false,
+            level: 1,
+            fields: 'id,path,displayName,children::isNotEmpty',
+        },
+    },
+}
+
+const AUTH_LOOKUP = {
+    APP: new Set(['ALL', 'M_dhis-web-user']),
+    USER: new Set([
+        'ALL',
+        'F_USER_VIEW',
+        'F_REPLICATE_USER',
+        'F_USER_ADD',
+        'F_USER_ADD_WITHIN_MANAGED_GROUP',
+        'F_USER_DELETE',
+        'F_USER_DELETE_WITHIN_MANAGED_GROUP',
+    ]),
+    GROUP: new Set([
+        'ALL',
+        'F_USERGROUP_DELETE',
+        'F_USERGROUP_PRIVATE_ADD',
+        'F_USERGROUP_PUBLIC_ADD',
+        'F_USER_GROUPS_READ_ONLY_ADD_MEMBERS',
+        'F_USERGROUP_MANAGING_RELATIONSHIPS_ADD',
+        'F_USERGROUP_MANAGING_RELATIONSHIPS_VIEW',
+    ]),
+    ROLE: new Set([
+        'ALL',
+        'F_USERROLE_DELETE',
+        'F_USERROLE_PRIVATE_ADD',
+        'F_USERROLE_PUBLIC_ADD',
+    ]),
+    USER_CREATE: new Set([
+        'ALL',
+        'F_USER_ADD',
+        'F_USER_ADD_WITHIN_MANAGED_GROUP',
+    ]),
+    GROUP_CREATE: new Set([
+        'ALL',
+        'F_USERGROUP_PRIVATE_ADD',
+        'F_USERGROUP_PUBLIC_ADD',
+        'F_USER_GROUPS_READ_ONLY_ADD_MEMBERS',
+        'F_USERGROUP_MANAGING_RELATIONSHIPS_ADD',
+    ]),
+    ROLE_CREATE: new Set([
+        'ALL',
+        'F_USERROLE_PRIVATE_ADD',
+        'F_USERROLE_PUBLIC_ADD',
+    ]),
+}
+
+const CurrentUserProvider = ({ children }) => {
+    const { data, error, fetching, loading, refetch } = useDataQuery(query)
+
+    if (loading || fetching) {
+        return (
+            <Layer translucent>
+                <CenteredContent>
+                    <CircularLoader />
+                </CenteredContent>
+            </Layer>
+        )
+    }
+
+    if (error) {
+        return (
+            <NoticeBox error title={i18n.t('Network error')}>
+                {i18n.t('Could not load current user information.')}
+            </NoticeBox>
+        )
+    }
+
+    const value = {
```

Comment:
By the time we get here, we can be assured nothing is undefined. There was no error and the data is not loading, so we must have data. The only reason for this to go wrong is if there is a problem or change in the backend, and I prefer stuff like that to blow up close to where it went wrong (i.e. the response handling) because this makes it a lot easier to debug. Usually trying to write resilient code in places like this just leads to other bugs in the rest of the app.